diff --git a/contracts/mocks/MulticallTokenMockUpgradeable.sol b/contracts/mocks/MulticallTokenMockUpgradeable.sol
index 7bd4ae88..7be0c4ff 100644
--- a/contracts/mocks/MulticallTokenMockUpgradeable.sol
+++ b/contracts/mocks/MulticallTokenMockUpgradeable.sol
@@ -8,7 +8,7 @@ import "../proxy/utils/Initializable.sol";
 
 contract MulticallTokenMockUpgradeable is Initializable, ERC20MockUpgradeable, MulticallUpgradeable {
     function __MulticallTokenMock_init(uint256 initialBalance) internal onlyInitializing {
-        __ERC20_init_unchained(name, symbol);
+        __ERC20_init_unchained("MulticallToken", "BCT");
         __ERC20Mock_init_unchained("MulticallToken", "BCT", msg.sender, initialBalance);
     }
 
