diff --git a/contracts/token/ERC20/extensions/ERC20VotesCompUpgradeable.sol b/contracts/token/ERC20/extensions/ERC20VotesCompUpgradeable.sol
index 6f1f8182..0f09ea48 100644
--- a/contracts/token/ERC20/extensions/ERC20VotesCompUpgradeable.sol
+++ b/contracts/token/ERC20/extensions/ERC20VotesCompUpgradeable.sol
@@ -25,13 +25,6 @@ import "../../../proxy/utils/Initializable.sol";
  * _Available since v4.2._
  */
 abstract contract ERC20VotesCompUpgradeable is Initializable, ERC20VotesUpgradeable {
-    function __ERC20VotesComp_init() internal onlyInitializing {
-        __Context_init_unchained();
-        __EIP712_init_unchained(name, "1");
-        __ERC20Votes_init_unchained();
-        __ERC20VotesComp_init_unchained();
-    }
-
     function __ERC20VotesComp_init_unchained() internal onlyInitializing {
     }
     /**
