diff --git a/contracts/mocks/wizard/MyGovernor1Upgradeable.sol b/contracts/mocks/wizard/MyGovernor1Upgradeable.sol
index 66662e9e..c53c6479 100644
--- a/contracts/mocks/wizard/MyGovernor1Upgradeable.sol
+++ b/contracts/mocks/wizard/MyGovernor1Upgradeable.sol
@@ -16,7 +16,7 @@ contract MyGovernor1Upgradeable is
     GovernorCountingSimpleUpgradeable
 {
     function __MyGovernor1_init(IVotesUpgradeable _token, TimelockControllerUpgradeable _timelock) internal onlyInitializing {
-        __EIP712_init_unchained(name_, version());
+        __EIP712_init_unchained("MyGovernor", version());
         __Governor_init_unchained("MyGovernor");
         __GovernorTimelockControl_init_unchained(_timelock);
         __GovernorVotes_init_unchained(_token);
diff --git a/contracts/mocks/wizard/MyGovernor2Upgradeable.sol b/contracts/mocks/wizard/MyGovernor2Upgradeable.sol
index 63bc024c..0f854226 100644
--- a/contracts/mocks/wizard/MyGovernor2Upgradeable.sol
+++ b/contracts/mocks/wizard/MyGovernor2Upgradeable.sol
@@ -18,7 +18,7 @@ contract MyGovernor2Upgradeable is
     GovernorCountingSimpleUpgradeable
 {
     function __MyGovernor2_init(IVotesUpgradeable _token, TimelockControllerUpgradeable _timelock) internal onlyInitializing {
-        __EIP712_init_unchained(name_, version());
+        __EIP712_init_unchained("MyGovernor", version());
         __Governor_init_unchained("MyGovernor");
         __GovernorTimelockControl_init_unchained(_timelock);
         __GovernorVotes_init_unchained(_token);
diff --git a/contracts/mocks/wizard/MyGovernor3Upgradeable.sol b/contracts/mocks/wizard/MyGovernor3Upgradeable.sol
index 27a8411c..2f59b8e5 100644
--- a/contracts/mocks/wizard/MyGovernor3Upgradeable.sol
+++ b/contracts/mocks/wizard/MyGovernor3Upgradeable.sol
@@ -16,7 +16,7 @@ contract MyGovernorUpgradeable is
     GovernorVotesQuorumFractionUpgradeable
 {
     function __MyGovernor_init(IVotesUpgradeable _token, TimelockControllerUpgradeable _timelock) internal onlyInitializing {
-        __EIP712_init_unchained(name_, version());
+        __EIP712_init_unchained("MyGovernor", version());
         __Governor_init_unchained("MyGovernor");
         __GovernorTimelockControl_init_unchained(_timelock);
         __GovernorVotes_init_unchained(_token);
