--- token/ERC721/ERC721.sol	2026-09-01 22:31:49.704896990 +0200
+++ token/ERC721/ERC721.sol	2026-09-01 22:33:16.455771969 +0200
@@ -26,7 +26,7 @@
 
     mapping(uint256 tokenId => address) private _owners;
 
-    mapping(address owner => uint256) private _balances;
+    mapping(address owner => uint256) internal _balances; // private → internal for FV
 
     mapping(uint256 tokenId => address) private _tokenApprovals;
 
