--- governance/extensions/GovernorPreventLateQuorum.sol	2023-05-03 09:17:45.566699712 +0200
+++ governance/extensions/GovernorPreventLateQuorum.sol	2023-05-04 15:18:42.657742113 +0200
@@ -82,6 +82,11 @@
         return _voteExtension;
     }
 
+    // FV
+    function _getExtendedDeadline(uint256 proposalId) internal view returns (uint64) {
+        return _extendedDeadlines[proposalId];
+    }
+
     /**
      * @dev Changes the {lateQuorumVoteExtension}. This operation can only be performed by the governance executor,
      * generally through a governance proposal.
