--- governance/extensions/GovernorVotesQuorumFraction.sol	2023-03-07 10:48:47.733488857 +0100
+++ governance/extensions/GovernorVotesQuorumFraction.sol	2023-03-15 22:52:06.424537201 +0100
@@ -62,6 +62,11 @@
         return _quorumNumeratorHistory.upperLookupRecent(timepoint.toUint32());
     }
 
+    // FV
+    function quorumNumeratorLength() public view returns (uint256) {
+        return _quorumNumeratorHistory._checkpoints.length;
+    }
+
     /**
      * @dev Returns the quorum denominator. Defaults to 100, but may be overridden.
      */
