--- governance/Governor.sol	2023-03-07 10:48:47.730155491 +0100
+++ governance/Governor.sol	2023-03-10 10:13:31.926616811 +0100
@@ -216,6 +216,16 @@
         return _proposals[proposalId].proposer;
     }
 
+    // FV
+    function _isExecuted(uint256 proposalId) internal view returns (bool) {
+        return _proposals[proposalId].executed;
+    }
+
+    // FV
+    function _isCanceled(uint256 proposalId) internal view returns (bool) {
+        return _proposals[proposalId].canceled;
+    }
+
     /**
      * @dev Amount of votes already cast passes the threshold limit.
      */
