--- governance/Governor.sol	2023-05-03 09:17:45.566699712 +0200
+++ governance/Governor.sol	2023-05-04 15:18:42.667741565 +0200
@@ -224,6 +224,21 @@
         return _proposals[proposalId].proposer;
     }
 
+    // FV
+    function _isExecuted(uint256 proposalId) internal view returns (bool) {
+        return _proposals[proposalId].executed;
+    }
+
+    // FV
+    function _isCanceled(uint256 proposalId) internal view returns (bool) {
+        return _proposals[proposalId].canceled;
+    }
+
+    // FV
+    function _governanceCallLength() internal view returns (uint256) {
+        return _governanceCall.length();
+    }
+
     /**
      * @dev Amount of votes already cast passes the threshold limit.
      */
