--- governance/extensions/GovernorPreventLateQuorum.sol	2023-03-15 17:13:06.879632860 +0100
+++ governance/extensions/GovernorPreventLateQuorum.sol	2023-03-15 14:14:59.121060484 +0100
@@ -84,6 +84,11 @@
         return _voteExtension;
     }
 
+    // FV
+    function _getExtendedDeadline(uint256 proposalId) internal view returns (uint64) {
+        return _extendedDeadlines[proposalId];
+    }
+
     /**
      * @dev Changes the {lateQuorumVoteExtension}. This operation can only be performed by the governance executor,
      * generally through a governance proposal.
