--- governance/TimelockController.sol	2023-08-11 20:33:54
+++ governance/TimelockController.sol	2023-08-13 10:07:34
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: MIT
 // OpenZeppelin Contracts (last updated v4.9.0) (governance/TimelockController.sol)
 
-pragma solidity ^0.8.20;
+pragma solidity ^0.8.19;
 
 import {AccessControl} from "../access/AccessControl.sol";
 import {ERC721Holder} from "../token/ERC721/utils/ERC721Holder.sol";
