--- token/ERC721/ERC721.sol	2023-08-11 20:33:54
+++ token/ERC721/ERC721.sol	2023-08-13 09:56:16
@@ -214,6 +214,11 @@
         }
     }
 
+    // FV
+    function _getApproved(uint256 tokenId) internal view returns (address) {
+        return _tokenApprovals[tokenId];
+    }
+
     /**
      * @dev Unsafe write access to the balances, used by extensions that "mint" tokens using an {ownerOf} override.
      *
