--- utils/math/SafeCast.sol	2023-08-11 20:33:54
+++ utils/math/SafeCast.sol	2023-08-13 09:56:16
@@ -2,7 +2,7 @@
 // OpenZeppelin Contracts (last updated v4.8.0) (utils/math/SafeCast.sol)
 // This file was procedurally generated from scripts/generate/templates/SafeCast.js.
 
-pragma solidity ^0.8.20;
+pragma solidity ^0.8.19;
 
 /**
  * @dev Wrappers over Solidity's uintXX/intXX casting operators with added overflow
