diff --git a/contracts/vendor/erc4337-entrypoint/core/EntryPoint.sol b/contracts/vendor/erc4337-entrypoint/core/EntryPoint.sol
index 778115b1..44501524 100644
--- a/contracts/vendor/erc4337-entrypoint/core/EntryPoint.sol
+++ b/contracts/vendor/erc4337-entrypoint/core/EntryPoint.sol
@@ -15,10 +15,8 @@ import "./Helpers.sol";
 import "./NonceManager.sol";
 import "./UserOperationLib.sol";

-// import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
-// import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
-import "../../../utils/introspection/ERC165.sol"; // OZ edit
-import "../../../utils/ReentrancyGuard.sol"; // OZ edit
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";

 /*
  * Account-Abstraction (EIP-4337) singleton EntryPoint implementation.
